### PR TITLE
makefiles/cflags.inc.mk: Add -Wcast-align

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

As the title says. This is useful to detect issues when doing things like

```C
uint8_t foo[4];
uint32_t *bar = (uint32_t *)foo;
*bar = 0xacdcacdc;
```

Because in the above example `foo` has an alignment requirement of 1 byte, but `bar` has (on most 32 bit archs) an alignment requirement of 4 bytes. Unaligned writes will (depending on architecture) either result faults, writes to the next aligned address instead (usually rounded always down), or slow performance (and writes no longer being atomically).

### Testing procedure

Green Murdock.

### Issues/PRs references

Depends on:

- [x] https://github.com/RIOT-OS/RIOT/pull/15056
- [x] https://github.com/RIOT-OS/RIOT/pull/15057
- [x] https://github.com/RIOT-OS/RIOT/pull/15059
- [x] https://github.com/RIOT-OS/RIOT/pull/15088
- [x] https://github.com/RIOT-OS/RIOT/pull/15099
- [x] https://github.com/RIOT-OS/RIOT/pull/15100
- [x] https://github.com/RIOT-OS/RIOT/pull/15463
- [x] https://github.com/RIOT-OS/RIOT/pull/15464
- [x] https://github.com/RIOT-OS/RIOT/pull/17149
- [x] https://github.com/RIOT-OS/RIOT/pull/17150
- [x] https://github.com/RIOT-OS/RIOT/pull/17151
- [x] https://github.com/RIOT-OS/RIOT/pull/17153
- [x] https://github.com/RIOT-OS/RIOT/pull/17156
- [x] https://github.com/RIOT-OS/RIOT/pull/17192
- [x] https://github.com/RIOT-OS/RIOT/pull/17194
- [x] https://github.com/RIOT-OS/RIOT/pull/17152
- [x] https://github.com/RIOT-OS/RIOT/pull/17154
- [x] https://github.com/RIOT-OS/RIOT/pull/17155
- [x] https://github.com/RIOT-OS/RIOT/pull/17157
- [x] https://github.com/RIOT-OS/RIOT/pull/17195

### CODE REVIEW

**Comment by maribu** [comment]
Let's see how many issues this uncovers.

**Comment by maribu** [comment]
> Let's see how many issues this uncovers.

Oh oh :-( Sounds like some nights of bug fixing.

**Comment by fjmolinas** [comment]
How many of these are there? Do we need to enable them gradually?

**Comment by maribu** [comment]
> benpicco added CI: ready for build and removed CI: ready for build labels 9 hours ago

I really should take a look into the PR website next time before I do a rebase to retrigger Murdock :-/

**Comment by maribu** [comment]
@bergzand can you take a look at [this log](https://ci.riot-os.org/RIOT-OS/RIOT/14955/daa4b619adfb9b2fccdbd2ed53e1c52a178933e2/output/compile/tests/driver_adcxx1c/weact-f401ce:gnu.txt)? I believe that this is not a false positive, but a bug. Can you confirm this?

**Comment by maribu** [comment]
Rebased on top of current master

**Comment by bergzand** [comment]
> @bergzand can you take a look at [this log](https://ci.riot-os.org/RIOT-OS/RIOT/14955/daa4b619adfb9b2fccdbd2ed53e1c52a178933e2/output/compile/tests/driver_adcxx1c/weact-f401ce:gnu.txt)? I believe that this is not a false positive, but a bug. Can you confirm this?

Both should not cause bugs and I would rather call it sloppy coding than bugs or false positives. I'll open a PR to fix them.

**Comment by bergzand** [comment]
@maribu Thanks for the added commit, looks good to me and looks similar to what I would have modified to fix the issue

**Comment by fjmolinas** [comment]
Last logs show:

```
can.c:152:18: error: cast increases required alignment of target type [-Werror=cast-align]
     can_t *dev = (can_t *)candev;
```

**Comment by maribu** [comment]
> Last logs show:

I'm aware. So far Murdock always reached the 500 failed builds leading to premature ending of the CI run within seconds. Fixing all those issues feels a bit like trying to boil the ocean :-/

**Comment by maribu** [comment]
I split all the real fixes out into their own PRs, so that they can be reviewed properly. (And also the quick extension of `make compile-commands`, that is very much unrelated.)

After they all have been merged and this is rebased, this PR should only contains commits that add `-Wno-cast-align` to a bunch of pkgs and that silences false positives.

**Comment by fjmolinas** [comment]
@maribu there is a whitespace issues, please squash that one in right away!

File: tests/pkg_utensor/main.cpp
Comment (fjmolinas) [inline tests/pkg_utensor/main.cpp:38]: Thre is a whitespace here

**Review by fjmolinas — APPROVED** [review]
ACK, lets not let this one diverge, please squash the whitespace issue

**Comment by maribu** [comment]
Done. I also reordered the commits to first fix compilation with `-Wcast-align` and only at the end add `-Wcast-align`, so that bisecting will be less pain in the ass.

**Comment by fjmolinas** [comment]
Oh no another one showed up @maribu :( can you squash that in?

**Comment by maribu** [comment]
Thanks for the review!

**Pull request merged**
